### ISSUE TITLE
Manually upgraded iina fails to play sound on different codec videos in the same instance

### ISSUE THREAD

**Issue by MarioMastr** [iina/iina#4789]
**System and IINA version:**
M1 Macbook Air, 8GB - iina 1.3.4; custom built with mpv 0.37.0 and ffmpeg 6.1.1

- macOS 
- 14.2.1
- IINA 
- iina 1.3.4; custom built with mpv 0.37.0 and ffmpeg 6.1.1

**Expected behavior:**
Every time a video plays the audio should just be able to play.

**Actual behavior:**
Sometimes it doesn't. Testing with the goal of trying to get a good log tells me that the issue stems from trying to open a video with a different codec to one you have already opened in that instance of iina. Switching the audio device restores audio, and so does restarting, though opening a video with a different codec after restarting will cause the issue to return.

<details>
<summary><b>mpv log:</b></summary>

[mpv.log](https://github.com/iina/iina/files/14028399/mpv.log)
[iina.log](https://github.com/iina/iina/files/14028398/iina.log)

</details>

<br>

**Steps to reproduce:**
1. Open iina.
2. Open a video.
3. Open a video with a different video codec.
4. The second video will not have any sound.

- [ ] MPV does not have this problem.

**How often does this happen?**
Basically every time.

**Comment by low-batt** [comment]
Thank you for reporting this. Reproduced for me. Also reproduced with the `mpv` `master` branch.

As `mpv` handles audio, it is very likely the root cause of this is in `mpv`.

Currently the team is focused on a bug fix release using the same libraries that are used in IINA 1.3.4. Once that release is done my expectation is that we will upgrade the libraries IINA is dependent upon. Looks like this problem will need to be investigated and solved before that can happen.

**Comment by krackers** [comment]
@low-batt How does iina use mpv when playing two videos? My understanding is that mpv itself only supports playing one video or audio track at a given time. So the only way to have two videos playing at once is essentially to have two separate mpv processes.

**Comment by low-batt** [comment]
I believe the actual mpv player restriction is that it only supports one window. You can open multiple videos and play them side by side in the one window.

The one window restriction is intentional as explained in [this comment](https://github.com/mpv-player/mpv/issues/8742#issuecomment-821980170) on https://github.com/mpv-player/mpv/issues/8742, a request to support multiple windows:
> duplicate of #1377, #6865 and probably some more. the app bundle isn't really meant to open several instances by itself. several windows are theoretically possible but are not within the scope of mpv itself (the core doesn't support this) but rather of libmpv users.
>
> if you really want that behaviour you have to write a wrapper script yourself with the open command. there are several examples of that in the first linked issue. alternatively you could use iina.

IINA calls [mpv_create](https://github.com/mpv-player/mpv/blob/dcf8adf28967cb138ece45d0f383d3e83f14f826/libmpv/client.h#L481) to create a core for each window.

I think I correctly identified the mpv commit that introduced the audio problem. See the mpv issue.

**Comment by krackers** [comment]
>You can open multiple videos and play them side by side in the one window.

I don't follow, this basically replaces the currently loaded file with the new one, so it's not side-by-side in the same window?

>IINA calls [mpv_create](https://github.com/mpv-player/mpv/blob/dcf8adf28967cb138ece45d0f383d3e83f14f826/libmpv/client.h#L481) to create a core for each window.

Yes this makes sense, so based on this the AO state should not be shared between different windows (it's a completely separate `mpctx`). I'm basically trying to figure out why adding this part 

```
for (int n = 0; n < buffer_list->mNumberBuffers; n++)
        buffer_list->mBuffers[n].mDataByteSize = samples * ao->sstride;
```

would cause issues. If they are completely separate processes (like launching 2 different mpv instances) there's clearly no issue, so why would having two separate core threads now cause issues? Maybe there is some threading assumption that CoreAudio has about only one active callback at a given time?

**Comment by low-batt** [comment]
On playing side-by-side in the same window see [this comment](https://github.com/iina/iina/issues/3964#issuecomment-1267503453) for a specific example of how to do that with IINA (also works with mpv). To do this you use a filter and the mpv [external-files](https://mpv.io/manual/stable/#options-external-files) option:
> **--external-files=\<file-list>**
>    Load a file and add all of its tracks. This is useful to play different files together (for example audio from one file, video from another), or for advanced `--lavfi-complex` used (like playing two video files at the same time).

On the question about the code shown, we are now into debugging mpv code, so that discussion should really occur in the mpv issue I entered: https://github.com/mpv-player/mpv/issues/13348

I don't know enough about mpv internals to answer the question, but an important key to the problem is that once the second video is playing without audio, if you at that point change the audio device to another set of speakers audio starts playing. So the problem has something to do with the two cores interacting with the same audio device.

**Comment by low-batt** [comment]
A fix for https://github.com/mpv-player/mpv/issues/13348 has been merged. I pulled the latest `mpv` master, built `libmpv` and tested playing multiple files with IINA. All working fine.

I am leaving this issue open as a reminder that IINA can not use `mpv` 0.38.0 as the fix was merged after the release was cut. To use 0.38.0 IINA will need to apply a small patch to `mpv`.

**Comment by MarioMastr** [comment]
Only just looking back through this issue. Nice to see that everything has been rather dealt with!

**Comment by low-batt** [comment]
@MarioMastr Thanks again for reporting this one.  We were not aware of the problem and it gave us a head start on obtaining a fix from mpv.

Closing this as the IINA dependencies have been updated to mpv 0.38.0 patched with the required fix.

If you build IINA be sure to download the latest dependencies using `other/download_libs.sh`. There was one more patch applied to IINA's `libmpv` build recently. Not sure if that has made it to the server yet.

With respects to audio there is a new `Audio driver` setting that supports switching from the `Core Audio` driver to the new `AVFoundation` driver.
